comp/core/tagger/taggerimpl/tagstore/entity_tags.go

```diff
@@ -147,3 +200,160 @@ func (e *EntityTags) shouldRemove() bool {
+	source             string
+	expiryDate         time.Time
+	standardTags       []string
+	cachedAll          tagset.HashedTags // Low + orchestrator + high
+	cachedOrchestrator tagset.HashedTags // Low + orchestrator (subslice of cachedAll)
+	cachedLow          tagset.HashedTags // Sub-slice of cachedAll
+	isExpired          bool
+}
+
+func newEntityTagsWithSingleSource(entityID string, source string) *EntityTagsWithSingleSource {
+	return &EntityTagsWithSingleSource{
+		entityID: entityID,
+		source:   source,
+	}
+}
+
+func (e *EntityTagsWithSingleSource) toEntity() types.Entity {
+	cachedAll := e.cachedAll.Get()
+	cachedOrchestrator := e.cachedOrchestrator.Get()
+	cachedLow := e.cachedLow.Get()
+
+	return types.Entity{
+		ID:           e.entityID,
+		StandardTags: e.getStandard(),
+		// cachedAll contains low, orchestrator and high cardinality tags, in this order.
+		// cachedOrchestrator and cachedLow are subslices of cachedAll, starting at index 0.
+		HighCardinalityTags:         cachedAll[len(cachedOrchestrator):],
+		OrchestratorCardinalityTags: cachedOrchestrator[len(cachedLow):],
+		LowCardinalityTags:          cachedLow,
+	}
+}
+
+func (e *EntityTagsWithSingleSource) getStandard() []string {
+	return e.standardTags
+}
+
+func (e *EntityTagsWithSingleSource) getHashedTags(cardinality types.TagCardinality) tagset.HashedTags {
+	switch cardinality {
+	case types.HighCardinality:
+		return e.cachedAll
+	case types.OrchestratorCardinality:
+		return e.cachedOrchestrator
+	default:
+		return e.cachedLow
+	}
+}
+
+func (e *EntityTagsWithSingleSource) tagsForSource(source string) *sourceTags {
+	if source != e.source {
+		return nil
+	}
+
+	return &sourceTags{
+		lowCardTags:          e.cachedLow.Get(),
+		orchestratorCardTags: e.cachedAll.Slice(e.cachedLow.Len(), e.cachedOrchestrator.Len()).Get(),
+		highCardTags:         e.cachedAll.Slice(e.cachedOrchestrator.Len(), e.cachedAll.Len()).Get(),
+		standardTags:         e.standardTags,
+		expiryDate:           e.expiryDate,
+	}
+}
+
+func (e *EntityTagsWithSingleSource) tagsBySource() map[string][]string {
+	return map[string][]string{e.source: e.cachedAll.Get()}
+}
+
+func (e *EntityTagsWithSingleSource) setTagsForSource(source string, tags sourceTags) {
+	if source != e.source {
+		return
+	}
+
+	e.standardTags = tags.standardTags
+
+	all := make([]string, 0, len(tags.lowCardTags)+len(tags.orchestratorCardTags)+len(tags.highCardTags))
+	all = append(all, tags.lowCardTags...)
+	all = append(all, tags.orchestratorCardTags...)
+	all = append(all, tags.highCardTags...)
+
+	cached := tagset.NewHashedTagsFromSlice(all)
+
+	e.cachedAll = cached
+	e.cachedLow = cached.Slice(0, len(tags.lowCardTags))
+	e.cachedOrchestrator = cached.Slice(0, len(tags.lowCardTags)+len(tags.orchestratorCardTags))
+}
+
+func (e *EntityTagsWithSingleSource) sources() []string {
+	return []string{e.source}
+}
+
+func (e *EntityTagsWithSingleSource) deleteSource(source string, expiryDate time.Time) {
```

Comment:
same question as before (for `EntityTagsWithMultipleSource`)

Reply:
Renamed in the new commit :+1: